src/input/src/lib.rs

```diff
@@ -114,7 +119,167 @@ pub enum Input {
     /// Window gained or lost cursor.
     Cursor(bool),
     /// Window closed.
-    Close
+    Close(CloseArgs),
+    /// Render graphics.
+    Render(RenderArgs),
+    /// After rendering and swapping buffers.
+    AfterRender(AfterRenderArgs),
+    /// Update the state of the application.
+    Update(UpdateArgs),
+    /// Do background tasks that can be done incrementally.
+    Idle(IdleArgs),
+    /// Custom event.
+    ///
+    /// When comparing two custom events for equality,
+    /// they always return `false`.
+    Custom(EventId, Arc<Any>),
```

Comment:
It seems the introduction of this variant has made it so that `Input` is no longer `Send` or `Sync`:

```
    |         ^ the trait `std::marker::Send` is not implemented for `std::any::Any + 'static`
    |
    = note: `std::any::Any + 'static` cannot be sent between threads safely
    = note: required because of the requirements on the impl of `std::marker::Send` for `std::sync::Arc<std::any::Any + 'static>`
    = note: required because it appears within the type `conrod::event::Input`
    = note: required by `std::marker::Send`
```
```
^ the trait `std::marker::Sync` is not implemented for `std::any::Any + 'static`
```

This has broken some downstream threaded conrod code, where a conrod thread was receiving `Input` events from a separate thread (currently adding a test case to avoid letting it happen again in the future).

 Perhaps this could be `Any + Send + Sync`?

Now that I think of it, conrod should probably just switch to using its own `Input` type, as it does not require any of these new variants that have been added (this is why we originally switched from using `piston::input::Event` to `piston::input::Input`, however it looks like this PR merges the two undoing our switch anyway).

I just noticed that you opened an issue for this a couple of days ago at #1150. I'm sorry that I missed it, I would have been in favour of keeping the two separate - I'm not quite sure I understand the reasoning behind the merger.